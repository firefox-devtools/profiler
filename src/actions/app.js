/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { oneLine } from 'common-tags';
import {
  getSelectedTab,
  getDataSource,
  getLocalTrackOrderByPid,
} from 'firefox-profiler/selectors/url-state';
import {
  getTrackThreadHeights,
  getIsEventDelayTracksEnabled,
  getIsExperimentalCPUGraphsEnabled,
  getIsExperimentalProcessCPUTracksEnabled,
} from 'firefox-profiler/selectors/app';
import {
  getLocalTracksByPid,
  getThreads,
  getCounters,
} from 'firefox-profiler/selectors/profile';
import { sendAnalytics } from 'firefox-profiler/utils/analytics';
import {
  stateFromLocation,
  withHistoryReplaceStateSync,
} from 'firefox-profiler/app-logic/url-handling';
import { finalizeProfileView } from './receive-profile';
import { fatalError } from './errors';
import {
  addEventDelayTracksForThreads,
  initializeLocalTrackOrderByPid,
  addProcessCPUTracksForProcess,
} from 'firefox-profiler/profile-logic/tracks';
import { selectedThreadSelectors } from 'firefox-profiler/selectors/per-thread';
import {
  getIsCPUUtilizationProvided,
  getAreThereAnyProcessCPUCounters,
} from 'firefox-profiler/selectors/cpu';

import type {
  Profile,
  ThreadsKey,
  CssPixels,
  Action,
  ThunkAction,
  UrlState,
  UploadedProfileInformation,
  IndexIntoCategoryList,
} from 'firefox-profiler/types';
import type { TabSlug } from 'firefox-profiler/app-logic/tabs-handling';
import type {
  BrowserConnection,
  BrowserConnectionStatus,
} from 'firefox-profiler/app-logic/browser-connection';

export function changeSelectedTab(selectedTab: TabSlug): ThunkAction<void> {
  return (dispatch, getState) => {
    const previousTab = getSelectedTab(getState());
    if (previousTab !== selectedTab) {
      sendAnalytics({
        hitType: 'pageview',
        page: selectedTab,
      });
      dispatch({
        type: 'CHANGE_SELECTED_TAB',
        selectedTab,
      });
    }
  };
}

export function changeProfilesToCompare(profiles: string[]): Action {
  return {
    type: 'CHANGE_PROFILES_TO_COMPARE',
    profiles,
  };
}

export function startFetchingProfiles(): Action {
  return { type: 'START_FETCHING_PROFILES' };
}

export function urlSetupDone(): ThunkAction<void> {
  return (dispatch, getState) => {
    dispatch({ type: 'URL_SETUP_DONE' });

    // After the url setup is done, we can successfully query our state about its
    // initial page.
    const dataSource = getDataSource(getState());
    sendAnalytics({
      hitType: 'pageview',
      page: dataSource === 'none' ? 'home' : getSelectedTab(getState()),
    });
    sendAnalytics({
      hitType: 'event',
      eventCategory: 'datasource',
      eventAction: dataSource,
    });
  };
}

export function show404(url: string): Action {
  return { type: 'ROUTE_NOT_FOUND', url };
}

export function changeSidebarOpenState(tab: TabSlug, isOpen: boolean): Action {
  return { type: 'CHANGE_SIDEBAR_OPEN_STATE', tab, isOpen };
}

export function invalidatePanelLayout(): Action {
  return { type: 'INCREMENT_PANEL_LAYOUT_GENERATION' };
}

/**
 * The viewport component provides a hint to use shift to zoom scroll. The first
 * time a user does this, the hint goes away.
 */
export function setHasZoomedViaMousewheel() {
  return { type: 'HAS_ZOOMED_VIA_MOUSEWHEEL' };
}

/**
 * This function is called when we start setting up the initial url state.
 * It takes the location and profile data, converts the location into url
 * state and then dispatches relevant actions to finalize the view.
 * `profile` parameter can be null when the data source can't provide the profile
 * and the url upgrader step is not needed (e.g. 'from-browser').
 */
export function setupInitialUrlState(
  location: Location,
  profile: Profile | null,
  browserConnection: BrowserConnection | null
): ThunkAction<void> {
  return (dispatch) => {
    let urlState;
    try {
      urlState = stateFromLocation(location, profile);
    } catch (e) {
      if (e.name === 'UrlUpgradeError') {
        // The error is an URL upgrade error, let's fire a fatal error.
        // If there's a service worker update, the class `ServiceWorkerManager`
        // will automatically reload in case the new code knows how to handle
        // this URL version.
        dispatch(fatalError(e));
        return;
      }
      // The location could not be parsed, show a 404 instead.
      console.error(e);
      dispatch(show404(location.pathname + location.search));
      return;
    }

    // Validate the initial URL state. We can't refresh on a from-file URL.
    if (urlState.dataSource === 'from-file') {
      urlState = null;
    }

    // Normally having multiple dispatches is an anti pattern, but here it's
    // necessary because we are doing different things inside those actions and
    // they can't be merged because we are also calling those seperately on
    // other parts of the code.
    // The first dispatch here updates the url state, then changes state as the url
    // setup is done, and lastly finalizes the profile view since everything is set up now.
    // All of this is done while the history is replaced, as this is part of the initial
    // load process.
    withHistoryReplaceStateSync(() => {
      dispatch(updateUrlState(urlState));
      dispatch(finalizeProfileView(browserConnection));
      dispatch(urlSetupDone());
    });
  };
}

/**
 * This function is called when a browser navigation event happens. A new UrlState
 * is generated when the window.location is serialized, or the state is pulled out of
 * the history API.
 */
export function updateUrlState(newUrlState: UrlState | null): Action {
  return { type: 'UPDATE_URL_STATE', newUrlState };
}

export function reportTrackThreadHeight(
  threadsKey: ThreadsKey,
  height: CssPixels
): ThunkAction<void> {
  return (dispatch, getState) => {
    const trackThreadHeights = getTrackThreadHeights(getState());
    const previousHeight = trackThreadHeights[threadsKey];
    if (previousHeight !== height) {
      // Guard against unnecessary dispatches. This could happen frequently.
      dispatch({
        type: 'UPDATE_TRACK_THREAD_HEIGHT',
        height,
        threadsKey,
      });
    }
  };
}

/**
 * This action dismisses the newly published state. This happens when a user first
 * uploads a profile. We only want to remember this when we fist open the profile.
 */
export function dismissNewlyPublished(): Action {
  return { type: 'DISMISS_NEWLY_PUBLISHED' };
}

/**
 * Called when a user has started dragging a file. Used for loading
 * profiles with the drag and drop component.
 */
export function startDragging(): Action {
  return { type: 'START_DRAGGING' };
}

/**
 * Called when a user has stopped dragging a file.
 */
export function stopDragging(): Action {
  return { type: 'STOP_DRAGGING' };
}

/**
 * Called when a custom drag and drop overlay is mounted. This lets
 * the app know that we shouldn't create a default overlay.
 */
export function registerDragAndDropOverlay(): Action {
  return { type: 'REGISTER_DRAG_AND_DROP_OVERLAY' };
}

/**
 * Called when a custom drag and drop overlay is unmounted.
 */
export function unregisterDragAndDropOverlay(): Action {
  return { type: 'UNREGISTER_DRAG_AND_DROP_OVERLAY' };
}

/*
 * This action enables the event delay tracks. They are hidden by default because
 * they are usually for power users and not so meaningful for average users.
 * There is no UI that triggers this action in the profiler interface. Instead,
 * users have to enable this from the developer console by writing this line:
 * `experimental.enableEventDelayTracks()`
 */
export function enableEventDelayTracks(): ThunkAction<boolean> {
  return (dispatch, getState) => {
    if (getIsEventDelayTracksEnabled(getState())) {
      console.error(
        'Tried to enable the event delay tracks, but they are already enabled.'
      );
      return false;
    }

    if (
      selectedThreadSelectors.getSamplesTable(getState()).eventDelay ===
      undefined
    ) {
      // Return early if the profile doesn't have eventDelay values.
      console.error(oneLine`
        Tried to enable the event delay tracks, but this profile does
        not have eventDelay values. It is likely an older profile.
      `);
      return false;
    }

    const oldLocalTracks = getLocalTracksByPid(getState());
    const localTracksByPid = addEventDelayTracksForThreads(
      getThreads(getState()),
      oldLocalTracks
    );
    const localTrackOrderByPid = initializeLocalTrackOrderByPid(
      getLocalTrackOrderByPid(getState()),
      localTracksByPid,
      null
    );
    dispatch({
      type: 'ENABLE_EVENT_DELAY_TRACKS',
      localTracksByPid,
      localTrackOrderByPid,
    });

    return true;
  };
}

/*
 * This action enables the CPU graph tracks. They are hidden by default because
 * they are usually for power users and not so meaningful for average users.
 * There is no UI that triggers this action in the profiler interface. Instead,
 * users have to enable this from the developer console by writing this line:
 * `experimental.enableCPUGraphs()`
 */
export function enableExperimentalCPUGraphs(): ThunkAction<boolean> {
  return (dispatch, getState) => {
    if (getIsExperimentalCPUGraphsEnabled(getState())) {
      console.error(
        'Tried to enable the CPU graph tracks, but they are already enabled.'
      );
      return false;
    }

    if (!getIsCPUUtilizationProvided(getState())) {
      // Return early if the profile doesn't have threadCPUDelta values.
      console.error(oneLine`
        Tried to enable the CPU graph tracks, but this profile does
        not have threadCPUDelta values. It is likely an older profile.
      `);
      return false;
    }

    dispatch({
      type: 'ENABLE_EXPERIMENTAL_CPU_GRAPHS',
    });

    return true;
  };
}

/*
 * This action enables the process CPU tracks. They are hidden by default because
 * the front-end work is not done for this data yet.
 * There is no UI that triggers this action in the profiler interface. Instead,
 * users have to enable this from the developer console by writing this line:
 * `experimental.enableExperimentalProcessCPUTracks()`
 */
export function enableExperimentalProcessCPUTracks(): ThunkAction<boolean> {
  return (dispatch, getState) => {
    if (getIsExperimentalProcessCPUTracksEnabled(getState())) {
      console.error(
        'Tried to enable the process CPU tracks, but they are already enabled.'
      );
      return false;
    }

    if (
      !getIsCPUUtilizationProvided(getState()) &&
      getAreThereAnyProcessCPUCounters(getState())
    ) {
      // Return early if the profile doesn't have threadCPUDelta values or
      // doesn't have any experimental process CPU counters.
      console.error(oneLine`
        Tried to enable the process CPU tracks, but this profile does
        not have threadCPUDelta values or process CPU threads.
      `);
      return false;
    }

    const oldLocalTracks = getLocalTracksByPid(getState());
    const localTracksByPid = addProcessCPUTracksForProcess(
      getCounters(getState()),
      oldLocalTracks
    );
    const localTrackOrderByPid = initializeLocalTrackOrderByPid(
      getLocalTrackOrderByPid(getState()),
      localTracksByPid,
      null
    );

    dispatch({
      type: 'ENABLE_EXPERIMENTAL_PROCESS_CPU_TRACKS',
      localTracksByPid,
      localTrackOrderByPid,
    });

    return true;
  };
}

/**
 * This caches the profile data in the local state for synchronous access.
 */
export function setCurrentProfileUploadedInformation(
  uploadedProfileInformation: UploadedProfileInformation | null
): Action {
  return {
    type: 'SET_CURRENT_PROFILE_UPLOADED_INFORMATION',
    uploadedProfileInformation,
  };
}

export function profileRemotelyDeleted(): Action {
  // Ideally we should store the current profile data in a local indexeddb, and
  // set the URL to /local/<indexeddb-key>.
  return { type: 'PROFILE_REMOTELY_DELETED' };
}

export function updateBrowserConnectionStatus(
  browserConnectionStatus: BrowserConnectionStatus
): Action {
  return { type: 'UPDATE_BROWSER_CONNECTION_STATUS', browserConnectionStatus };
}

export function toggleOpenCategoryInSidebar(
  kind: string,
  category: IndexIntoCategoryList
): Action {
  return {
    type: 'TOGGLE_SIDEBAR_OPEN_CATEGORY',
    kind,
    category,
  };
}
