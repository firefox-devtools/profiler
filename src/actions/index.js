/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import * as icons from './icons';
import * as profileView from './profile-view';
import * as publish from './publish';
import * as receiveProfile from './receive-profile';
import * as zippedProfiles from './zipped-profiles';
import * as code from './code';

export default Object.assign(
  {},
  app,
  icons,
  profileView,
  publish,
  receiveProfile,
  zippedProfiles,
  code
);
