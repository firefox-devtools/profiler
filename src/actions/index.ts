/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
import * as app from './app';
import * as icons from './icons';
import * as l10n from './l10n';
import * as profileView from './profile-view';
import * as publish from './publish';
import * as receiveProfile from './receive-profile';
import * as zippedProfiles from './zipped-profiles';
import * as code from './code';

export default Object.assign(
  {},
  app,
  icons,
  l10n,
  profileView,
  publish,
  receiveProfile,
  zippedProfiles,
  code
);
