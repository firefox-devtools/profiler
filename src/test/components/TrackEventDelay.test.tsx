/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
import { Provider } from 'react-redux';
import { fireEvent } from '@testing-library/react';

import { render } from 'firefox-profiler/test/fixtures/testing-library';
import { TrackEventDelay } from '../../components/timeline/TrackEventDelay';
import { ensureExists } from '../../utils/types';
import { enableEventDelayTracks } from '../../actions/app';

import {
  autoMockCanvasContext,
  flushDrawLog,
} from '../fixtures/mocks/canvas-context';
import { mockRaf } from '../fixtures/mocks/request-animation-frame';
import { storeWithProfile } from '../fixtures/stores';
import {
  addRootOverlayElement,
  removeRootOverlayElement,
  getMouseEvent,
} from '../fixtures/utils';
import { getProfileWithEventDelays } from '../fixtures/profiles/processed-profile';
import { autoMockElementSize } from '../fixtures/mocks/element-size';

import type { IndexIntoSamplesTable } from 'firefox-profiler/types';
import type { CssPixels } from '../../types/units';

// The following constants determine the size of the drawn graph.
const SAMPLE_COUNT = 8;
const PIXELS_PER_SAMPLE = 10;
const GRAPH_WIDTH = PIXELS_PER_SAMPLE * SAMPLE_COUNT;
const GRAPH_HEIGHT = 10;

function getSamplesPixelPosition(
  sampleIndex: IndexIntoSamplesTable,
  samplePosition: number
): CssPixels {
  // Compute the pixel position of the center of a given sample.
  return sampleIndex * PIXELS_PER_SAMPLE + PIXELS_PER_SAMPLE * samplePosition;
}

/**
 * This test verifies that the event delay track can draw a graph of the responsiveness.
 */
describe('TrackEventDelay', function () {
  function setup(isTrackEnabled: boolean = true) {
    const profile = getProfileWithEventDelays();
    const store = storeWithProfile(profile);
    const { getState, dispatch } = store;
    const flushRafCalls = mockRaf();

    // Enable the event delay tracks if we want to see them, they are disabled by default
    if (isTrackEnabled) {
      dispatch(enableEventDelayTracks());
    }

    const renderResult = render(
      <Provider store={store}>
        <TrackEventDelay threadIndex={0} />
      </Provider>
    );
    const { container } = renderResult;

    // WithSize uses requestAnimationFrame
    flushRafCalls();

    const canvas = ensureExists(
      container.querySelector('.timelineTrackEventDelayCanvas'),
      `Couldn't find the event delay canvas, with selector .timelineTrackEventDelayCanvas`
    );
    const getTooltipContents = () =>
      document.querySelector('.timelineTrackEventDelayTooltip');
    const getEventDelayDot = () =>
      container.querySelector('.timelineTrackEventDelayGraphDot');
    const moveMouseAtEventDelay = (index: number, pos: number) =>
      fireEvent(
        canvas,
        getMouseEvent('mousemove', {
          pageX: getSamplesPixelPosition(index, pos),
        })
      );

    return {
      ...renderResult,
      dispatch,
      getState,
      profile,
      store,
      canvas,
      getTooltipContents,
      moveMouseAtEventDelay,
      flushRafCalls,
      getEventDelayDot,
    };
  }

  autoMockCanvasContext();
  autoMockElementSize({ width: GRAPH_WIDTH, height: GRAPH_HEIGHT });
  beforeEach(addRootOverlayElement);
  afterEach(removeRootOverlayElement);

  it('matches the component snapshot when event delay tracks are disabled', () => {
    const { container } = setup(false);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches the 2d canvas draw snapshot', () => {
    const { flushRafCalls } = setup();
    flushRafCalls();
    expect(flushDrawLog()).toMatchSnapshot();
  });

  it('can create a tooltip', function () {
    const { moveMouseAtEventDelay, getTooltipContents, canvas } = setup();
    expect(getTooltipContents()).toBeFalsy();
    moveMouseAtEventDelay(1, 0.5);
    expect(getTooltipContents()).toBeTruthy();
    fireEvent.mouseLeave(canvas);
    expect(getTooltipContents()).toBeFalsy();
  });

  it('has a tooltip that matches the snapshot', function () {
    const { moveMouseAtEventDelay, getTooltipContents } = setup();
    moveMouseAtEventDelay(5, 0.5);
    expect(getTooltipContents()).toMatchSnapshot();
  });

  it('draws a dot on the graph', function () {
    const { moveMouseAtEventDelay, getEventDelayDot } = setup();
    expect(getEventDelayDot()).toBeFalsy();
    moveMouseAtEventDelay(1, 0.5);
    expect(getEventDelayDot()).toBeTruthy();
  });

  it('can draw a dot on both extremes of the graph', function () {
    const { moveMouseAtEventDelay, getEventDelayDot } = setup();
    expect(getEventDelayDot()).toBeFalsy();
    moveMouseAtEventDelay(0, 0.25);
    expect(getEventDelayDot()).toBeTruthy();
    moveMouseAtEventDelay(8, 0);
    expect(getEventDelayDot()).toBeTruthy();
  });

  it('draws a dot that matches the snapshot', function () {
    const { moveMouseAtEventDelay, getEventDelayDot } = setup();
    moveMouseAtEventDelay(1, 0.5);
    expect(getEventDelayDot()).toMatchSnapshot();
  });
});
