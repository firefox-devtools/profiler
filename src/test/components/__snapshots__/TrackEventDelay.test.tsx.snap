// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TrackEventDelay draws a dot that matches the snapshot 1`] = `
<div
  class="timelineTrackEventDelayGraphDot"
  style="left: 14.933333333333334px; top: 39.327586206896555px;"
/>
`;

exports[`TrackEventDelay has a tooltip that matches the snapshot 1`] = `
<div
  class="timelineTrackEventDelayTooltip"
>
  <div
    class="timelineTrackEventDelayTooltipLine"
  >
    <span
      class="timelineTrackEventDelayTooltipNumber"
    >
      37ms
    </span>
     event delay
  </div>
  <div
    class="timelineTrackEventDelayTooltipLine"
  >
    <span
      class="timelineTrackEventDelayTooltipNumber"
    >
      58ms
    </span>
     delay range in graph
  </div>
</div>
`;

exports[`TrackEventDelay matches the 2d canvas draw snapshot 1`] = `
Array [
  Array [
    "clearRect",
    0,
    0,
    80,
    40,
  ],
  Array [
    "set lineWidth",
    2,
  ],
  Array [
    "set lineJoin",
    "bevel",
  ],
  Array [
    "set strokeStyle",
    "rgba(255, 0, 57, 0.7)",
  ],
  Array [
    "set fillStyle",
    "rgba(255, 0, 57, 0.3)",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    0,
    39,
  ],
  Array [
    "lineTo",
    0.5333333333333333,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    1.0666666666666667,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    1.6,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    2.1333333333333333,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    2.6666666666666665,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    3.2,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    3.7333333333333334,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    4.266666666666667,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    4.8,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    5.333333333333333,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    5.866666666666666,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    6.4,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    6.933333333333334,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    7.466666666666667,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    8,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    8.533333333333333,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    9.066666666666666,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    9.6,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    10.133333333333333,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    10.666666666666666,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    11.2,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    11.733333333333333,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    12.266666666666666,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    12.8,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    13.333333333333334,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    13.866666666666667,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    14.4,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    14.933333333333334,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    15.466666666666667,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    16,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    16.53333333333333,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    17.066666666666666,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    17.6,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    18.133333333333333,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    18.666666666666668,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    19.2,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    19.733333333333334,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    20.266666666666666,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    20.8,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    21.333333333333332,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    21.866666666666667,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    22.4,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    22.933333333333334,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    23.466666666666665,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    24,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    24.53333333333333,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    25.066666666666666,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    25.6,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    26.133333333333333,
    38.3448275862069,
  ],
  Array [
    "lineTo",
    26.666666666666668,
    19.344827586206897,
  ],
  Array [
    "lineTo",
    27.2,
    0.3448275862068968,
  ],
  Array [
    "lineTo",
    27.733333333333334,
    1,
  ],
  Array [
    "lineTo",
    28.266666666666666,
    1.6551724137931032,
  ],
  Array [
    "lineTo",
    28.8,
    2.3103448275862064,
  ],
  Array [
    "lineTo",
    29.333333333333332,
    2.9655172413793096,
  ],
  Array [
    "lineTo",
    29.866666666666667,
    3.620689655172413,
  ],
  Array [
    "lineTo",
    30.4,
    4.275862068965516,
  ],
  Array [
    "lineTo",
    30.933333333333334,
    4.931034482758619,
  ],
  Array [
    "lineTo",
    31.466666666666665,
    5.586206896551722,
  ],
  Array [
    "lineTo",
    32,
    6.241379310344833,
  ],
  Array [
    "lineTo",
    32.53333333333333,
    6.896551724137929,
  ],
  Array [
    "lineTo",
    33.06666666666666,
    7.551724137931036,
  ],
  Array [
    "lineTo",
    33.6,
    8.206896551724139,
  ],
  Array [
    "lineTo",
    34.13333333333333,
    8.862068965517238,
  ],
  Array [
    "lineTo",
    34.666666666666664,
    9.517241379310345,
  ],
  Array [
    "lineTo",
    35.2,
    10.172413793103448,
  ],
  Array [
    "lineTo",
    35.733333333333334,
    10.827586206896552,
  ],
  Array [
    "lineTo",
    36.266666666666666,
    11.482758620689655,
  ],
  Array [
    "lineTo",
    36.8,
    12.137931034482762,
  ],
  Array [
    "lineTo",
    37.333333333333336,
    12.793103448275861,
  ],
  Array [
    "lineTo",
    37.86666666666667,
    13.448275862068964,
  ],
  Array [
    "lineTo",
    38.4,
    14.103448275862071,
  ],
  Array [
    "lineTo",
    38.93333333333333,
    14.75862068965517,
  ],
  Array [
    "lineTo",
    39.46666666666667,
    15.413793103448274,
  ],
  Array [
    "lineTo",
    40,
    16.06896551724138,
  ],
  Array [
    "lineTo",
    40.53333333333333,
    16.724137931034484,
  ],
  Array [
    "lineTo",
    41.06666666666666,
    17.379310344827584,
  ],
  Array [
    "lineTo",
    41.6,
    18.03448275862069,
  ],
  Array [
    "lineTo",
    42.13333333333333,
    18.689655172413794,
  ],
  Array [
    "lineTo",
    42.666666666666664,
    19.344827586206897,
  ],
  Array [
    "lineTo",
    43.2,
    0.3448275862068968,
  ],
  Array [
    "lineTo",
    43.733333333333334,
    1,
  ],
  Array [
    "lineTo",
    44.266666666666666,
    1.6551724137931032,
  ],
  Array [
    "lineTo",
    44.8,
    2.3103448275862064,
  ],
  Array [
    "lineTo",
    45.333333333333336,
    2.9655172413793096,
  ],
  Array [
    "lineTo",
    45.86666666666667,
    3.620689655172413,
  ],
  Array [
    "lineTo",
    46.4,
    4.275862068965516,
  ],
  Array [
    "lineTo",
    46.93333333333333,
    4.931034482758619,
  ],
  Array [
    "lineTo",
    47.46666666666667,
    5.586206896551722,
  ],
  Array [
    "lineTo",
    48,
    6.241379310344833,
  ],
  Array [
    "lineTo",
    48.53333333333333,
    6.896551724137929,
  ],
  Array [
    "lineTo",
    49.06666666666666,
    7.551724137931036,
  ],
  Array [
    "lineTo",
    49.6,
    8.206896551724139,
  ],
  Array [
    "lineTo",
    50.13333333333333,
    8.862068965517238,
  ],
  Array [
    "lineTo",
    50.666666666666664,
    9.517241379310345,
  ],
  Array [
    "lineTo",
    51.2,
    10.172413793103448,
  ],
  Array [
    "lineTo",
    51.733333333333334,
    10.827586206896552,
  ],
  Array [
    "lineTo",
    52.266666666666666,
    11.482758620689655,
  ],
  Array [
    "lineTo",
    52.8,
    12.137931034482762,
  ],
  Array [
    "lineTo",
    53.333333333333336,
    12.793103448275861,
  ],
  Array [
    "lineTo",
    53.86666666666667,
    13.448275862068964,
  ],
  Array [
    "lineTo",
    54.4,
    14.103448275862071,
  ],
  Array [
    "lineTo",
    54.93333333333333,
    14.75862068965517,
  ],
  Array [
    "lineTo",
    55.46666666666667,
    15.413793103448274,
  ],
  Array [
    "lineTo",
    56,
    16.06896551724138,
  ],
  Array [
    "lineTo",
    56.53333333333333,
    16.724137931034484,
  ],
  Array [
    "lineTo",
    57.06666666666666,
    17.379310344827584,
  ],
  Array [
    "lineTo",
    57.6,
    18.03448275862069,
  ],
  Array [
    "lineTo",
    58.13333333333333,
    18.689655172413794,
  ],
  Array [
    "lineTo",
    58.666666666666664,
    19.344827586206897,
  ],
  Array [
    "lineTo",
    59.199999999999996,
    13.448275862068964,
  ],
  Array [
    "lineTo",
    59.733333333333334,
    14.103448275862071,
  ],
  Array [
    "lineTo",
    60.266666666666666,
    14.75862068965517,
  ],
  Array [
    "lineTo",
    60.8,
    15.413793103448274,
  ],
  Array [
    "lineTo",
    61.333333333333336,
    16.06896551724138,
  ],
  Array [
    "lineTo",
    61.86666666666667,
    16.724137931034484,
  ],
  Array [
    "lineTo",
    62.4,
    17.379310344827584,
  ],
  Array [
    "lineTo",
    62.93333333333333,
    18.03448275862069,
  ],
  Array [
    "lineTo",
    63.46666666666667,
    18.689655172413794,
  ],
  Array [
    "lineTo",
    64,
    19.344827586206897,
  ],
  Array [
    "lineTo",
    64.53333333333333,
    20,
  ],
  Array [
    "lineTo",
    65.06666666666666,
    20.655172413793103,
  ],
  Array [
    "lineTo",
    65.6,
    21.310344827586206,
  ],
  Array [
    "lineTo",
    66.13333333333333,
    21.96551724137931,
  ],
  Array [
    "lineTo",
    66.66666666666667,
    22.620689655172416,
  ],
  Array [
    "lineTo",
    67.2,
    23.275862068965516,
  ],
  Array [
    "lineTo",
    67.73333333333333,
    23.93103448275862,
  ],
  Array [
    "lineTo",
    68.26666666666667,
    24.586206896551722,
  ],
  Array [
    "lineTo",
    68.8,
    25.241379310344826,
  ],
  Array [
    "lineTo",
    69.33333333333333,
    25.89655172413793,
  ],
  Array [
    "lineTo",
    69.86666666666666,
    26.551724137931036,
  ],
  Array [
    "lineTo",
    70.4,
    27.206896551724135,
  ],
  Array [
    "lineTo",
    70.93333333333334,
    27.862068965517242,
  ],
  Array [
    "lineTo",
    71.46666666666667,
    28.517241379310345,
  ],
  Array [
    "lineTo",
    72,
    29.17241379310345,
  ],
  Array [
    "lineTo",
    72.53333333333333,
    29.82758620689655,
  ],
  Array [
    "lineTo",
    73.06666666666666,
    30.482758620689655,
  ],
  Array [
    "lineTo",
    73.6,
    31.137931034482758,
  ],
  Array [
    "lineTo",
    74.13333333333333,
    31.79310344827586,
  ],
  Array [
    "lineTo",
    74.66666666666667,
    32.44827586206897,
  ],
  Array [
    "lineTo",
    75.2,
    39,
  ],
  Array [
    "lineTo",
    75.73333333333333,
    39,
  ],
  Array [
    "lineTo",
    76.26666666666667,
    39,
  ],
  Array [
    "lineTo",
    76.8,
    39,
  ],
  Array [
    "lineTo",
    77.33333333333333,
    39,
  ],
  Array [
    "lineTo",
    77.86666666666666,
    39,
  ],
  Array [
    "lineTo",
    78.4,
    39,
  ],
  Array [
    "lineTo",
    78.93333333333334,
    39,
  ],
  Array [
    "lineTo",
    79.46666666666667,
    39,
  ],
  Array [
    "lineTo",
    80,
    39,
  ],
  Array [
    "stroke",
  ],
  Array [
    "lineTo",
    80,
    40,
  ],
  Array [
    "lineTo",
    1,
    40,
  ],
  Array [
    "fill",
  ],
]
`;

exports[`TrackEventDelay matches the component snapshot when event delay tracks are disabled 1`] = `
<div
  class="timelineTrackEventDelay"
  style="height: 40px; --graph-height: 40px; --markers-height: 0px;"
>
  <div
    class="timelineTrackEventDelayGraph"
  >
    <canvas
      class="timelineTrackEventDelayCanvas"
      height="40"
      width="80"
    />
    <div
      class="timelineEmptyThreadIndicator"
    />
  </div>
</div>
`;
